File: internal/proxy/metrics.go
package proxy

import "sync/atomic"

type Metrics struct {
	totalRequests      atomic.Uint64
	manifestRequests   atomic.Uint64
	blobRequests       atomic.Uint64
	cacheHits          atomic.Uint64
	cacheMisses        atomic.Uint64
	upstreamRequests   atomic.Uint64
	upstreamFailures   atomic.Uint64
	cacheWriteFailures atomic.Uint64
}

type MetricsSnapshot struct {
	TotalRequests      uint64
	ManifestRequests   uint64
	BlobRequests       uint64
	CacheHits          uint64
	CacheMisses        uint64
	UpstreamRequests   uint64
	UpstreamFailures   uint64
	CacheWriteFailures uint64
}

func NewMetrics() *Metrics {
	return &Metrics{}
}

func (m *Metrics) Snapshot() MetricsSnapshot {
	return MetricsSnapshot{
		TotalRequests:      m.totalRequests.Load(),
		ManifestRequests:   m.manifestRequests.Load(),
		BlobRequests:       m.blobRequests.Load(),
		CacheHits:          m.cacheHits.Load(),
		CacheMisses:        m.cacheMisses.Load(),
		UpstreamRequests:   m.upstreamRequests.Load(),
		UpstreamFailures:   m.upstreamFailures.Load(),
		CacheWriteFailures: m.cacheWriteFailures.Load(),
	}
}

File: frontend/src/lib/api/types/metrics.ts
export interface PullInfo {
  host: string
  port: string
  virtualTag: string
  hostConfigured: boolean
}

export interface ProxyMetricsCounters {
  totalRequests: number
  manifestRequests: number
  blobRequests: number
  cacheHits: number
  cacheMisses: number
  upstreamRequests: number
  upstreamFailures: number
  cacheWriteFailures: number
}

export type ProxyMetricsWindow = 'today' | 'last7d' | 'last30d'

export interface ProxyMetrics {
  generatedAt: string
  windows: Record<ProxyMetricsWindow, ProxyMetricsCounters>
  cache: { usedBytes: number; maxBytes: number } | null
}
